File: sdk/javascript/Service.ts
/**
 * An IoT service state
 */
export default class Service {
  /**
   * @param name friendly name of the IoT service
   * @param deviceId identity of the device to which the IoT service belongs
   * @param organizationId identity of the organization to which the IoT service belongs
   * @param version version number of the IoT service
   * @param description a brief summary of the service's functions
   * @param lastUpdateTime the latest time that the service state has been updated
   */
  constructor(
    public name: string,
    public deviceId: string,
    public organizationId: string,
    public version: number = 0,
    public description: string = '',
    public lastUpdateTime: Date = new Date(0),
  ) {}

  /**
   * Get components that compose the service key
   *
   * @returns components that compose the service key
   */
  getKeyComponents() {
    return [this.organizationId, this.deviceId, this.name];
  }

  /**
   * Transform current service to a plain object
   *
   * @returns a plain object containing the service information
   */
  toObject(): { [key: string]: any } {
    return {
      name: this.name,
      deviceId: this.deviceId,
      organizationId: this.organizationId,
      version: this.version,
      description: this.description,
      lastUpdateTime: this.lastUpdateTime,
    };
  }

  /**
   * Transform current service to JSON string
   *
   * @returns JSON representation of the service
   */
  serialize() {
    return JSON.stringify(this.toObject());
  }

  /**
   * Check if the IoT service properties are valid
   *
   * @throws error when service fields are invalid
   */
  validate() {
    if (this.name === '') {
      throw new Error('missing service name in service definition');
    }
    if (this.deviceId === '') {
      throw new Error('missing device ID in service definition');
    }
    if (this.organizationId === '') {
      throw new Error('missing organization ID in service definition');
    }
    if (this.version === 0) {
      throw new Error('missing service version in service definition');
    }
    if (this.version !== (this.version | 0) || this.version < 0) {
      throw new Error('service version must be a positive integer');
    }
    if (this.lastUpdateTime.getTime() === 0) {
      throw new Error('missing service last update time in service definition');
    }
  }

  /**
   * Create a new device instance from an object
   *
   * @param obj an object that contains the device information
   * @returns a new device instance
   */
  static fromObject(obj: { [key: string]: any }): Service {
    return new Service(
      obj.name,
      obj.deviceId,
      obj.organizationId,
      obj.version ? obj.version : 0,
      obj.description ? obj.description : '',
      obj.lastUpdateTime ? new Date(obj.lastUpdateTime) : new Date(0),
    );
  }

  /**
   * Create an IoT service instance from its JSON representation
   *
   * @param data JSON string representing a service
   * @returns a new service instance
   */
  static deserialize(data: string) {
    return Service.fromObject(JSON.parse(data));
  }
}

File: common/service.go
package common

import (
	"encoding/json"
	"fmt"
	"time"
)

// Service an IoT service state
type Service struct {
	// Name name of the IoT service
	Name string `json:"name"`

	// DeviceId identity of the device to which the IoT service belongs
	DeviceId string `json:"deviceId"`

	// OrganizationId identity of the organization to which the IoT service belongs
	OrganizationId string `json:"organizationId"`

	// Version version number of the IoT service
	Version int32 `json:"version"`

	// Description a brief summary of the service's functions
	Description string `json:"description"`

	// LastUpdateTime the latest time that the service state has been updated
	LastUpdateTime time.Time `json:"lastUpdateTime"`
}

// GetKeyComponents return components that compose the IoT service key
func (s *Service) GetKeyComponents() []string {
	return []string{s.OrganizationId, s.DeviceId, s.Name}
}

// Serialize transform current IoT service to JSON string
func (s *Service) Serialize() ([]byte, error) {
	return json.Marshal(s)
}

// Validate check if the IoT service properties are valid
func (s *Service) Validate() error {
	if s.Name == "" {
		return fmt.Errorf("missing service name in service definition")
	}
	if s.DeviceId == "" {
		return fmt.Errorf("missing device ID in service definition")
	}
	if s.OrganizationId == "" {
		return fmt.Errorf("missing organization ID in service definition")
	}
	if s.Version == 0 {
		return fmt.Errorf("missing service version in service definition")
	}
	if s.Version < 0 {
		return fmt.Errorf("service version must be a positive integer")
	}
	if s.LastUpdateTime.IsZero() {
		return fmt.Errorf("missing service last update time in device definition")
	}

	return nil
}

// DeserializeService create an IoT service instance from its JSON representation
func DeserializeService(data []byte) (*Service, error) {
	service := new(Service)

	if err := json.Unmarshal(data, service); err != nil {
		return nil, err
	}

	return service, nil
}
